File: app/src/main/java/com/example/ign_app/ArticleAdapter.java
package com.example.ign_app;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.ign_app.CommentPackage.FeedComment;
import com.example.ign_app.CommentPackage.content.ContentComment;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ArticleAdapter extends  RecyclerView.Adapter<ArticleAdapter.ArticleViewHolder> {


    private Context mContext;
    private List<ArticleData> articleDataList;
    private static final String BASE_URL = "https://ign-apis.herokuapp.com/";
    List<CommentData> commentDataList;
    ArrayList<ContentComment> contentCommentArrayList;
    private static final int VIEW_TYPE_LOADING = 0;
    private static final int VIEW_TYPE_NORMAL = 1;
    private boolean isLoaderVisible = false;




    public ArticleAdapter(Context mContext, List<ArticleData> articleDataList) {
        this.mContext = mContext;
        this.articleDataList = articleDataList;
    //    this.commentDataList = commentDataList;

    }

    @NonNull
    @Override
    public ArticleViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        LayoutInflater inflater = LayoutInflater.from(mContext);
        View view = inflater.inflate(R.layout.articlecard, null);
        return new ArticleViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ArticleViewHolder holder, int position) {

        final ArticleData articleData = articleDataList.get(position);
        holder.head_txt.setText(articleData.getHeadline());
        holder.des_txt.setText(articleData.getDescription());
        holder.authorName_txt.setText(articleData.getAuthorName());
        Picasso.with(mContext).load(articleData.getUrlImage()).into(holder.article_image);


        final Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        CommentApi CApi = retrofit.create(CommentApi.class);
            Call<FeedComment> call2 = CApi.getCommentCount(articleData.getContentId());
            call2.enqueue(new Callback<FeedComment>() {
                @Override
                public void onResponse(Call<FeedComment> call, Response<FeedComment> response) {
                    contentCommentArrayList= response.body().getContent();

                    holder.comments.setText(contentCommentArrayList.get(0).getCount());

                }

                @Override
                public void onFailure(Call<FeedComment> call, Throwable t) {

                }
            });

        if (articleData.getAuthorImage().isEmpty()) {
            Picasso.with(mContext).load(R.drawable.ic_launcher_background).into(holder.author_image);
        } else{
            Picasso.with(mContext).load(articleData.getAuthorImage()).into(holder.author_image);
        }

        holder.relativeLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(mContext, WebViewActivity.class);
                i.putExtra("URL", articleData.getSlug());
                mContext.startActivity(i);

            }
        });
    }


    @Override
    public int getItemCount() {
        return articleDataList.size();
    }



        public void addItems(List<ArticleData> articleData) {
        articleDataList.addAll(articleData);
        notifyDataSetChanged();
    }

    class ArticleViewHolder extends RecyclerView.ViewHolder{

        TextView head_txt, des_txt, authorName_txt, comments;
        ImageView article_image;
        CircleImageView author_image;
        RelativeLayout relativeLayout;
        WebView webView;
        public ArticleViewHolder(@NonNull View itemView) {
            super(itemView);
            head_txt = itemView.findViewById(R.id.headline_article);
            des_txt = itemView.findViewById(R.id.article_description);
            authorName_txt = itemView.findViewById(R.id.article_author_name);
            article_image = itemView.findViewById(R.id.article_image);
            author_image = itemView.findViewById(R.id.article_author);
            comments = itemView.findViewById(R.id.comments);
            relativeLayout = itemView.findViewById(R.id.relative_layout);

        }
    }
}

File: app/src/main/java/com/example/ign_app/adapter/ArticleAdapter.kt
package com.example.ign_app.adapter

import android.content.Context
import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.ign_app.R
import com.example.ign_app.activities.ArticleTab
import com.example.ign_app.activities.WebViewActivity
import com.example.ign_app.commentpackage.CommentApiResponse
import com.example.ign_app.commentpackage.content.ContentComment
import com.example.ign_app.helper.ArticleData
import com.example.ign_app.networkcalls.CommentApi
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.articlecard.view.*
import org.jetbrains.anko.doAsync
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import com.example.ign_app.adapter.ArticleAdapter.ArticleViewHolder

import retrofit2.converter.gson.GsonConverterFactory
import java.util.*

 class ArticleAdapter(val mContext: Context, private val articleDataList: MutableList<ArticleData>) :
        RecyclerView.Adapter<ArticleViewHolder>() {
    var contentCommentArrayList: ArrayList<ContentComment>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArticleViewHolder {
        val inflater = LayoutInflater.from(mContext)
        val view = inflater.inflate(R.layout.articlecard, null)
        return ArticleViewHolder(view)
    }

    override fun onBindViewHolder(holder: ArticleViewHolder, position: Int) {
        if(holder!=null) {
            holder.card.headline_article.text = articleDataList[position].headline
            holder.card.article_description.text = articleDataList[position].description
            holder.card.article_author_name.text = articleDataList[position].authorName
            Picasso.with(mContext).load(articleDataList[position].urlImage).into(holder.card.article_image)

            val retrofit = Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()

            val CApi = retrofit.create(CommentApi::class.java)
            val call2 = CApi.getCommentCount(articleDataList[position].contentId)

            doAsync {
                call2!!.enqueue(object : Callback<CommentApiResponse> {
                    override fun onResponse(call: Call<CommentApiResponse>, response: Response<CommentApiResponse>) {
                        contentCommentArrayList = response.body()!!.content
                        holder.card.comments.text = contentCommentArrayList!![0].count
                        Log.d("countcomment", contentCommentArrayList!![0].count)
                    }

                    override fun onFailure(call: Call<CommentApiResponse>, t: Throwable) {
                        Log.d("Failed", t.message)
                    }
                })
            }

            if (articleDataList[position].authorImage.isEmpty()) {
                Picasso.with(mContext).load(R.drawable.ic_launcher_background).into(holder.card.article_image)
            } else {
                Picasso.with(mContext).load(articleDataList[position].authorImage).into(holder.card.article_author)
            }
            holder.card.relative_layout.setOnClickListener {
                val i = Intent(mContext, WebViewActivity::class.java)
                i.putExtra("URL", articleDataList[position].slug)
                mContext.startActivity(i)
            }
        }
        else{
            Log.d("ITS NULL","NULL")
        }
    }

    override fun getItemCount(): Int {
        return articleDataList!!.size
    }

    fun addItems(articleData: List<ArticleData>) {
        articleDataList?.addAll(articleData)
        notifyDataSetChanged()
    }

    class ArticleViewHolder(val card : View) : RecyclerView.ViewHolder(card)


    companion object {
        private const val BASE_URL = "https://ign-apis.herokuapp.com/"
    }

}




